cargo-marker/src/cli.rs

```diff
@@ -0,0 +1,63 @@
+use clap::{builder::ValueParser, Arg, ArgAction, Command};
+
+use crate::VERSION;
+
+const AFTER_HELP_MSG: &str = r#"CARGO ARGS
+    All arguments after double dashes(`--`) will be passed to cargo.
+    These options are the same as for `cargo check`.
+
+EXAMPLES:
+    * `cargo marker -l ./marker_lints`
+"#;
+
+pub fn get_clap_config() -> Command {
+    Command::new(VERSION)
+        .arg(
+            Arg::new("version")
+                .short('V')
+                .long("version")
+                .action(ArgAction::SetTrue)
+                .help("Print version info and exit"),
+        )
+        .arg(
+            Arg::new("verbose")
+                .short('v')
+                .long("verbose")
+                .action(ArgAction::SetTrue)
+                .help("Print additional debug information to the console"),
+        )
+        .arg(
+            Arg::new("test-setup")
+                .long("test-setup")
+                .action(ArgAction::SetTrue)
+                .help("This flag will compile the lint crate and print all relevant environment values"),
+        )
+        .subcommand(setup_command())
+        .subcommand(check_command())
+        .args(check_command_args())
+        .after_help(AFTER_HELP_MSG)
+        .override_usage("cargo-marker [OPTIONS] -- <CARGO ARGS>")
+}
+
+fn setup_command() -> Command {
+    Command::new("setup")
+        .about("A collection of commands to setup marker")
+        .after_help("By default this will install the driver for rustc.")
+}
+
+fn check_command() -> Command {
+    Command::new("check")
+        .about("Run marker on a local package")
+        .args(check_command_args())
+}
+
+fn check_command_args() -> impl IntoIterator<Item = impl Into<Arg>> {
```

Comment:
Hey, not to pile up on an already huge pile of work you're doing, but this `impl` return type, causes `marker_driver_rustc` to panic, when trying to lint with default test lint in `marker_lints`.

Seems like it's because of [this `TODO` in here](https://github.com/rust-marker/marker/blob/8d6c683cf80af105a1d5d4a7ca464a2a297b0e24/marker_driver_rustc/src/conversion/ty.rs#L88), just saying that you might want to get that fixed before merging this :)

<details>
<summary>
Panic
</summary>

```
thread '<unnamed>' panicked at 'not yet implemented: Ty {
    hir_id: HirId {
        owner: OwnerId {
            def_id: DefId(0:24 ~ cargo_marker[a207]::cli::check_command_args),
        },
        local_id: 37,
    },
    kind: OpaqueDef(
        ItemId {
            owner_id: OwnerId {
                def_id: DefId(0:165 ~ cargo_marker[a207]::cli::check_command_args::{opaque#0}),
            },
        },
        [],
        false,
    ),
    span: cargo-marker/src/cli.rs:54:28: 54:68 (#0),
}', marker_driver_rustc/src/conversion/ty.rs:88:17
```

</details>

Reply:
Hey @Niki4tap, nice to see you in this repo. Welcome to marker!

> not to pile up on an already huge pile of work you're doing

No worries at all. I appreciate reviews, it might take a bit more work, but that's better than fixing the code later again. It's also helpful to discuss alternative solutions.

The `impl Trait` type slipped my mind when I worked on the rustc backend. #55 includes a change in rustc that effects how `impl Trait` is represented. I waited for the nightly version until removing the Todo. Then life happened, and it's already the end of December ^^.

I might merge this PR before the nightly bump to avoid conflicts, but then I'll try to fix it in that PR :)

Reply:
> Welcome to marker!

Thanks!

> I waited for the nightly version until removing the Todo. Then life happened, and it's already the end of December ^^.

Yeah, that tends to happen

> I might merge this PR before the nightly bump to avoid conflicts, but then I'll try to fix it in that PR :)

No problem, just making sure you're aware of the issue :)